src/main/java/org/jitsi/videobridge/AbstractEndpoint.java

```diff
@@ -478,4 +476,45 @@ public void decrementSelectedCount()
     {
         // No-op
     }
+
+    /**
+     * Recreates this {@link AbstractEndpoint}'s media stream tracks based
+     * on the sources (and source groups) described in it's video channel.
+     */
+    public void recreateMediaStreamTracks()
+    {
+        ChannelShim videoChannel = getChannelOfMediaType(MediaType.VIDEO);
+        if (videoChannel != null)
+        {
+            MediaStreamTrackDesc[] tracks =
+                    MediaStreamTrackFactory.createMediaStreamTracks(
+                            videoChannel.getSources(),
+                            videoChannel.getSourceGroups());
+            setMediaStreamTracks(tracks);
+        }
+    }
+
+    /**
+     * Gets this {@link AbstractEndpoint}'s channel of media type
+     * {@code mediaType} (although it's not strictly enforced, endpoints have
+     * at most one channel with a given media type).
+     *
+     * @param mediaType the media type of the channel.
+     *
+     * @return the channel.
+     */
+    private ChannelShim getChannelOfMediaType(MediaType mediaType)
+    {
+        return
+            channelShims.stream()
+                    .filter(Objects::nonNull)
```

Comment:
i think this filter be after the weak reference `get`?

Reply:
EDIT: nevermind, i see you do the null check differently there, but could just do another `Objects::nonNull` filter

Reply:
You mean two filters instead of one filter with an "&&" in the condition?

Reply:
yeah but don't care either way actually